gowin/arch.cc

```diff
@@ -2245,6 +2312,54 @@ void Arch::fix_pll_nets(Context *ctx)
     }
 }
 
+// mark with hclk is used
+void Arch::mark_used_hclk(Context *ctx)
+{
+    for (auto &cell : ctx->cells) {
+        CellInfo *ci = cell.second.get();
+        if (ci->type != id_IOLOGIC) {
+            continue;
+        }
+        if (ci->attrs.count(id_IOLOGIC_FCLK)) {
+            continue;
+        }
+        // if it's an aux cell
+        if (ci->attrs.count(id_IOLOGIC_MASTER_CELL)) {
+            continue;
+        }
+        ci->setAttr(id_IOLOGIC_FCLK, Property("UNKNOWN"));
+
+        // *** FCLK
+        if (port_used(ci, id_FCLK)) {
+            NetInfo const *net = ci->getPort(id_FCLK);
+            for (auto const &user : net->users) {
+                if (user.cell != ci) {
+                    continue;
+                }
+                if (user.port != id_FCLK) {
+                    continue;
+                }
+                WireId dstWire = ctx->getNetinfoSinkWire(net, user, 0);
+                if (ctx->verbose) {
+                    log_info("   Cell:%s, port:%s, wire:%s\n", user.cell->name.c_str(this), user.port.c_str(this),
+                             dstWire.c_str(this));
+                }
+                for (PipId const &pip : getPipsUphill(dstWire)) {
```

Comment:
why the `const &`?

Reply:
why not just `PipId`? it's small enough that passing by value will be faster than passing by reference?